### ISSUE TITLE
Add FKE (ACL 2025)

### ISSUE THREAD

**Issue by vkehfdl1** [NomaDamas/AutoRAG-Research#193]
https://aclanthology.org/2025.findings-acl.522.pdf

plug&play

**Comment by vkehfdl1** [comment]
I'll implement the FKE (Fine-grained Knowledge Enhancement) pipeline for this issue.

## Implementation Plan

Based on the paper (ACL 2025 Findings), FKE is a **generation pipeline** with two key components:
1. **Disentangled Chain-of-Thought prompting** — extracts fine-grained sentence-level knowledge from retrieved documents
2. **Knowledge-enhanced generation** — uses the extracted fine-grained knowledge to generate more accurate answers

This maps cleanly to the existing `BaseGenerationPipeline` pattern (similar to `BasicRAGPipeline` but with a 2-step generation process like `ET2RAGPipeline`).

### Files to create/modify:
- `autorag_research/pipelines/generation/fke.py` — Pipeline + Config
- `configs/pipelines/generation/fke.yaml` — YAML config
- `autorag_research/pipelines/generation/__init__.py` — Export new classes
- `tests/autorag_research/pipelines/generation/test_fke_pipeline.py` — Tests

Starting implementation now.

**Comment by vkehfdl1** [comment]
## Implementation Plan for FKE Pipeline

Based on review of the paper ([ACL 2025 Findings](https://aclanthology.org/2025.findings-acl.522/)) and the existing codebase patterns.

### What is FKE?
FKE (Fine-grained Knowledge Enhancement) is a **plug-and-play generation pipeline** that enhances standard RAG by adding a knowledge extraction step before answer generation:
1. **Retrieve** documents for a query (standard retrieval pipeline)
2. **Extract fine-grained knowledge** — use CoT prompting to identify key factual sentences from retrieved chunks (LLM call #1)
3. **Generate answer** — use both original context AND extracted knowledge to produce a more accurate answer (LLM call #2)

This is a 2-step generation pipeline, similar in structure to `ET2RAGPipeline`.

### Files to Create
1. **`autorag_research/pipelines/generation/fke.py`** — Pipeline + Config classes
2. **`configs/pipelines/generation/fke.yaml`** — YAML configuration
3. **`tests/autorag_research/pipelines/generation/test_fke_pipeline.py`** — Unit tests

### File to Modify
4. **`autorag_research/pipelines/generation/__init__.py`** — Add FKE exports

### Implementation Details

#### `FKEPipelineConfig` (dataclass)
- Extends `BaseGenerationPipelineConfig`
- Fields:
  - `knowledge_extraction_prompt_template: str` — CoT prompt with `{context}` and `{query}` placeholders
  - `generation_prompt_template: str` — final generation prompt with `{context}`, `{knowledge}`, and `{query}` placeholders
  - `extraction_max_tokens: int | None = None`

#### `FKEPipeline` class
- Extends `BaseGenerationPipeline`
- `_generate(query_id, top_k)` algorithm:
  1. Retrieve chunks via `self._retrieval_pipeline._retrieve_by_id(query_id, top_k)`
  2. Fetch contents via `self._service.get_chunk_contents(chunk_ids)`
  3. Get query text via `self._service.get_query_text(query_id)`
  4. Build numbered context string (`[Document 1]`, `[Document 2]`, ...)
  5. **LLM Call #1** — Knowledge extraction: format extraction prompt → `await self._llm.ainvoke(prompt)`
  6. **LLM Call #2** — Answer generation: format generation prompt with context + extracted knowledge → `await self._llm.ainvoke(prompt)`
  7. Track token usage with `TokenUsageTracker` (from `autorag_research.util`)
  8. Return `GenerationResult(text, token_usage=tracker.total, metadata={chunk_ids, scores, extracted_knowledge})`

#### Default Prompts
**Knowledge Extraction (CoT):**
```
You are a knowledge extraction assistant. Given a query and retrieved documents,
extract the most relevant fine-grained factual information as a numbered list of key sentences.

Focus on:
- Specific facts, numbers, dates, and names relevant to the query
- Causal relationships and explanations
- Definitions and key concepts

Query: {query}

Retrieved Documents:
{context}

Extract the key fine-grained knowledge points relevant to answering the query:
```

**Answer Generation:**
```
Answer the following question using the provided context and extracted knowledge points.

Context:
{context}

Extracted Knowledge:
{knowledge}

Question: {query}

Answer:
```

#### YAML Config (`configs/pipelines/generation/fke.yaml`)
```yaml
_target_: autorag_research.pipelines.generation.fke.FKEPipelineConfig
description: "Fine-grained Knowledge Enhancement RAG (ACL 2025)"
name: fke
retrieval_pipeline_name: bm25
llm: gpt-4o-mini
top_k: 10
batch_size: 128
max_concurrency: 8
max_retries: 3
retry_delay: 1.0
```

#### Tests
- `test_fke_pipeline_config` — config returns correct class and kwargs
- `test_fke_generate_basic` — mock retrieval + service, verify 2 LLM calls (extraction + generation)
- `test_fke_generate_empty_retrieval` — graceful handling when no chunks
- `test_fke_generate_custom_prompts` — custom templates are used
- `test_fke_pipeline_config_dict` — `_get_pipeline_config()` returns correct dict
- All tests use `FakeListLLM` / mocks, no real API calls

### Conventions
- `TokenUsageTracker` for token tracking (no duplication)
- `logger = logging.getLogger("AutoRAG-Research")`
- Python 3.10+ type hints, 120 char line length
- `make check` must pass (ruff, ty, deptry)

### Status
Branch `feature/issue-193-fke-pipeline` created. Implementation blocked on Codex CLI availability — will execute as soon as tooling is ready.

**Comment by vkehfdl1** [comment]
Thanks for opening this issue. I ran an initial PM triage.

## Initial Assessment
- **Scope**: Medium — add FKE (ACL 2025) as a new pipeline. Noted as "plug&play".
- **Risk**: Low — additive pipeline implementation
- **Missing context**: Only a paper link provided. Need: (1) pipeline type (retrieval/generation), (2) key algorithmic steps, (3) what "plug&play" means in the context of this project — does it map to an existing pipeline pattern?
- **Implementation readiness**: `needs-details`

## Next Step
- Please summarize FKE's approach, identify the pipeline type, and list the key steps to implement. If "plug&play" means it follows an existing pattern closely, please specify which one.

**Comment by vkehfdl1** [comment]
- [x] **AI-understood issue summary**
  FKE (*Fine-grained Knowledge Enhancement for Retrieval-Augmented Generation*, Findings ACL 2025) is not a brand-new retriever by itself; it is a **plug-and-play enhancement layer** for an existing RAG pipeline. The paper adds two main pieces: (1) a two-step prompt that extracts **query-focused sentence-level evidence** from already retrieved documents, and (2) a **comparative decoding** step that blends document-level and fine-grained sentence-level generation signals. In the paper, this improves both multi-hop and single-hop QA over the base pipelines they augment.

- [x] **Why this issue is needed**
  This fits the repo direction well: AutoRAG-Research already implements paper-aligned retrieval/generation pipelines, and FKE is explicitly framed as a **plug-and-play upgrade** rather than a full retraining recipe. It also covers a gap that the current built-ins do not address directly: **sentence-level evidence extraction + generation-time control** on top of an existing retrieval pipeline. If implemented well, it could become a reusable enhancement pattern for `BasicRAG`-style workflows. The paper also reports meaningful gains over its bases: roughly **+2.8/+3.2 EM/F1 on 2Wiki**, **+4.9/+3.5 on HotpotQA**, **+6.3 Acc on PopQA**, and **+5.4 Acc on ARC**.

- [x] **Why this issue may not be needed**
  The risky part is that the paper’s strongest contribution is the **decoding-time probability fusion**, and the current codebase mostly treats the LLM as a generic LangChain generator (`ainvoke`) rather than as a full token-logit interface. There is some logprob support in the repo (for example, MAIN-RAG-style judging), but a **paper-faithful FKE implementation likely needs stronger/provider-specific logit access or a carefully defined approximation**. Also, the paper states that fine-grained knowledge / source code are available **by request via email**, so reproducibility may be weaker than for papers with public code. If we only want a practical sentence-extraction add-on, a simpler enhancement may be enough; if we want faithful FKE, this is a higher-effort item.

- [x] **Expected Outcome**
  A good outcome would be:
  1. a new **generation pipeline (or enhancement wrapper)** that composes with an existing retrieval pipeline,
  2. configurable fine-grained sentence extraction prompts,
  3. a clearly documented **"faithful if logits/logprobs are available, fallback if not"** behavior,
  4. YAML examples + docs,
  5. tests with mocked retrieval/LLM responses proving extraction, fusion/fallback logic, and executor integration.

- [x] **Concise implementation plan**
  1. **Scope the target first**: decide whether this issue means **paper-faithful FKE** or a **practical FKE-inspired variant** for the current abstractions.
  2. **Start from generation composition patterns already in the repo** (`BaseGenerationPipeline`, `BasicRAG`, `MAIN-RAG`) instead of inventing a new framework path.
  3. Implement the **fine-grained retrieval stage**: retrieve top-N docs from the injected retrieval pipeline, then use the existing LLM to extract query-focused sentences with the disentangled prompt.
  4. Implement **decoding enhancement** behind a capability boundary:
     - faithful path when the model/provider exposes usable logprobs/logits,
     - documented fallback path when it does not.
  5. Add **config + tests + docs** (mocked LLM/retrieval tests first, then one paper-style config example).
  6. Only after that, consider benchmarking against a baseline pipeline to confirm the feature is worth keeping.

- [x] **Agent Signature**
